[Security] Added docs about the authenticator-based Security system (#13704)
symfony/symfony-docs: security/experimental_authenticators.rst

Using the new Authenticator-based System
========================================

.. versionadded:: 5.1

    The authenticator-based system was introduced as
    :doc:`experimental feature </contributing/code/experimental>` in
    Symfony 5.1.

In Symfony 5.1, a new authentication system was introduced. This system
changes the internals of the Security system, to make it more extensible

> and easier to understand.

Reviewer: What about: more understandable
As imo, the code for an external pov can be not this easy :)

``enable_authenticator_manager`` setting:

.. configuration-block::

    .. code-block:: yaml

        # config/packages/security.yaml
        security:
            enable_authenticator_manager: true
            # ...

    .. code-block:: xml

        <!-- config/packages/security.xml -->
        <?xml version="1.0" encoding="UTF-8"?>
        <srv:container xmlns="http://symfony.com/schema/dic/security"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xmlns:srv="http://symfony.com/schema/dic/services"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config enable-authenticator-manager="true">
                <!-- ... -->
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        $container->loadFromExtension('security', [
            'enable_authenticator_manager' => true,
            // ...
        ]);

> The new system is backwards compatible with the current authentication

Author: We should maybe come up with a consistent name for the new system. "Guards" were very nice to use in a sentence. "Authenticator-based system" is both too technical and too long imho. "New system" on the other hand is too vague.

Reviewer: If you use the word « new » maybe use the word « old » instead of « current » ?
Otherwise, what about « next » instead of new?
As i understood, it will be the new when this will be the only way, isnt it?

Author: It's the new system, but it's experimental so it's not really the "current" system. Decided to keep using "new" or "authenticator-based system" - at least for now.

    ``entry_point: App\Security\CustomEntryPoint``)

.. _authenticators-removed-authentication-providers:

Creating a Custom Authenticator
-------------------------------

Security traditionally could be extended by writing
:doc:`custom authentication providers </security/custom_authentication_provider>`.
The authenticator-based system dropped support for these providers and
introduced a new authenticator interface as a base for custom
authentication methods.

.. tip::

    :doc:`Guard authenticators </security/guard_authentication>` are still
    supported in the authenticator-based system. It is however recommended
    to also update these when you're refactoring your application to the
    new system.

    The new authenticator interface has quite some overlap with the guard
    authenticator interface, making the rewrite easier. To avoid duplicated
    class names, it is recommended to remove the ``symfony/security-guard``
    package before rewriting your Security classes:

    .. code-block:: terminal

        $ composer remove symfony/security-guard

Authenticators should implement the
:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.

> The easiest way to use it is by extending

Reviewer: What about « An alternative could be to extends the abstract ... »

Because an abstract is not always easier imo, sometimes implementing fully is easier :)

:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AbstractAuthenticator`::

    // src/Security/ApiKeyAuthenticator.php
    namespace App\Security;

    use App\Entity\User;
    use Doctrine\ORM\EntityManagerInterface;
    use Symfony\Component\HttpFoundation\JsonResponse;
    use Symfony\Component\HttpFoundation\Request;
    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
    use Symfony\Component\Security\Core\Exception\AuthenticationException;
    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
    use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
    use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;

>     class ApiKeyAuthenticator extends AbstractAuthenticator

Author: I'm not sure I like having an API token authenticator, as it doesn't use the default `Passport` and has no success response. Doing a login form isn't great either (users should extend `AbstractFormLoginAuthenticator` then).
Does anyone have an idea for a good example authenticator that we can use?

Reviewer: What about a magic link auth?
Retrieves and checks data from request (ie a user uniq token)
If ok, a response that redirects to a default authenticated page

Author: I really liked this idea, until I realized this also used the `SelfValidatingPassport`. In the end, I think we can keep the current example for now. Let's see if it's confusing.

            function ($credentials, UserInterface $user) {
                return $user->getApiToken() === $credentials;
            },

            // The custom credentials
            $apiToken
        ));

.. note::

    If you don't need any credentials to be checked (e.g. a JWT token), you
    can use the
    :class:`Symfony\\Component\\Security\\Http\\Authenticator\\Passport\\SelfValidatingPassport`.
    This class only requires a user and optionally `Passport Badges`_.

Passport Badges
~~~~~~~~~~~~~~~

The ``Passport`` also optionally allows you to add *security badges*.
Badges attach more data to the passport (to extend security). By default,
the following badges are supported:

:class:`Symfony\\Component\\Security\\Http\\Authenticator\\Passport\\Badge\\RememberMeBadge`
    When this badge is added to the passport, the authenticator indicates
    remember me is supported. Whether remember me is actually used depends
    on special ``remember_me`` configuration. Read
    :doc:`/security/remember_me` for more information.

:class:`Symfony\\Component\\Security\\Http\\Authenticator\\Passport\\Badge\\PasswordUpgradeBadge`

>     This badge requires the plaintext password and a password upgrader.

Reviewer: I would write first the aim, then the requirement like you did for others

    Indicates that this user was pre-authenticated (i.e. before Symfony was
    initiated). This skips the
    :doc:`pre-authentication user checker </security/user_checkers>`.

For instance, if you want to add CSRF and password migration to your custom
authenticator, you would initialize the passport like this::

    // ...
    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
    use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
    use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;

    class LoginAuthenticator extends AbstractAuthenticator
    {
        public function authenticate(Request $request): PassportInterface
        {
            $password = $request->request->get('password');
            $username = $request->request->get('username');
            $csrfToken = $request->request->get('csrf_token');

            // ... get the $user from the $username and validate no
            // parameter is empty

>             return new Password($user, new PasswordCredentials($password), [

Reviewer 2: I guess it should be `new Passport`?

> Using the new Authenticator-based System

Reviewer 3: "system" looks too generic to me ... what about `Using the new Authenticator-based Security` ?

==========================================

.. versionadded:: 5.1

>     Authenticator-based security was introduced as

Reviewer 4: ```suggestion
    Authenticator-based security was introduced as an
```

                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config enable-authenticator-manager="true">
                <!-- ... -->

                <access-control>
                    <!-- allow unauthenticated users to access the login form -->
                    <rule path="^/admin/login" role="PUBLIC_ACCESS"/>

                    <!-- but require authentication for all other admin routes -->
                    <rule path="^/admin" role="ROLE_ADMIN"/>
                </access-control>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Component\Security\Http\Firewall\AccessListener;

        $container->loadFromExtension('security', [
            'enable_authenticator_manager' => true,

            // ...
            'access_control' => [
                // allow unauthenticated users to access the login form
                ['path' => '^/admin/login', 'roles' => AccessListener::PUBLIC_ACCESS],

                // but require authentication for all other admin routes

>                 ['path' => '^/admin', 'roles' => AccessListener::PUBLIC_ACCESS],

Reviewer 4: The role is wrong here

            ],
        ]);

.. _authenticators-required-entry-point:

Configuring the Authentication Entry Point
------------------------------------------

Sometimes, one firewall has multiple ways to authenticate (e.g. both a form
login and an API token authentication). In these cases, it is now required
to configure the *authentication entry point*. The entry point is used to

> generate a response when the user is not yet authenticated. This can be

Reviewer 4: ```suggestion
generate a response when the user is not yet authenticated but tries to access a page that
requires authentication. This can be
```

> used to for instance redirect the user to the login page.

Reviewer 3: `used to for instance redirect` -> `used for instance to redirect` ?

> You can configure this using the ``entry_point`` config setting:

Reviewer 3: `config setting` -> `option` ?

                <firewall name="main"
                    entry-point="form_login"
                >
                    <!-- allow authentication using a form or HTTP basic -->
                    <form-login/>
                    <http-basic/>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Component\Security\Http\Firewall\AccessListener;

        $container->loadFromExtension('security', [
            'enable_authenticator_manager' => true,

            // ...
            'firewalls' => [
                'main' => [
                    // allow authentication using a form or HTTP basic
                    'form_login' => null,
                    'http_basic' => null,

                    // configure the form authentication as the entry point for unauthenticated users
                    'entry_point' => 'form_login'
                ],
            ],
        ]);

.. note::

    You can also create your own authentication entry point by creating a
    class that implements
    :class:`Symfony\\Component\\Security\\Http\\EntryPoint\\AuthenticationEntryPointInterface`.

>     You can then set ``entry_point`` to the FQCN (e.g.

Reviewer 4: Hmm, I'm wondering if FCQN should say "service id" to be more accurate. But we would still show a class name in the example below.

    ``entry_point: App\Security\CustomEntryPoint``)

.. _authenticators-removed-authentication-providers:

Creating a Custom Authenticator
-------------------------------

Security traditionally could be extended by writing
:doc:`custom authentication providers </security/custom_authentication_provider>`.
The authenticator-based system dropped support for these providers and
introduced a new authenticator interface as a base for custom
authentication methods.

.. tip::

    :doc:`Guard authenticators </security/guard_authentication>` are still
    supported in the authenticator-based system. It is however recommended
    to also update these when you're refactoring your application to the
    new system.

>     The new authenticator interface has quite some overlap with the guard

Reviewer 4: ```suggestion
    The new authenticator interface has many similarities with the guard
```

    authenticator interface, making the rewrite easier. To avoid duplicated
    class names, it is recommended to remove the ``symfony/security-guard``
    package before rewriting your Security classes:

    .. code-block:: terminal

>         $ composer remove symfony/security-guard

Reviewer 4: Will that work if you only have, for example, `symfony/security-bundle` in your `composer.json`? That is the most likely use case, as `composer require security` gives you that package.

Author: Oh yes sorry, really good catch. This should be removed, as sadly `symfony/security-bundle` depends on the guard package.

Authenticators should implement the
:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.

> You can also

Reviewer 5: ```suggestion
You can also extend
```

    use Symfony\Component\HttpFoundation\Response;
    use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
    use Symfony\Component\Security\Core\Exception\AuthenticationException;
    use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
    use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
    use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
    use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;

    class ApiKeyAuthenticator extends AbstractAuthenticator
    {
        private $entityManager;

        public function __construct(EntityManagerInterface $entityManager)
        {
            $this->entityManager = $entityManager;
        }

        /**
         * Called on every request to decide if this authenticator should be
         * used for the request. Returning `false` will cause this authenticator
         * to be skipped.
         *
         * Returning `null` indicates this authenticator supports to be called

>          * lazily.

Reviewer 4: Even for me, I'm not sure what to do with this information :p. Is it relevant for users? If so, what does it *really* mean? When should I return null?

Author: It can be very relevant, but let's remove it and discuss it in a cache/lazy firewall related section instead (if we ever write one).

            }

            $user = $this->entityManager->getRepository(User::class)
                ->findOneBy(['apiToken' => $apiToken])
            ;
            if (null === $user) {
                throw new UsernameNotFoundException();
            }

            return new SelfValidatingPassport($user);
        }

        public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
        {
            // on success, let the request continue
            return null;
        }

        public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
        {
            $data = [
                // you may want to customize or obfuscate the message first
                'message' => strtr($exception->getMessageKey(), $exception->getMessageData())

                // or to translate this message
                // $this->translator->trans($exception->getMessageKey(), $exception->getMessageData())
            ];

            return new JsonResponse($data, Response::HTTP_UNAUTHORIZED);
        }
    }

The ``authenticate()`` method is the most important method of the

> authenticator. It's job is to extract credentials (e.g. username &

Reviewer 3: `It's job` -> `Its job`

password, or API tokens) from the ``Request`` object and transform these
into a security
:class:`Symfony\\Component\\Security\\Http\\Authenticator\\Passport\\Passport`.

Security Passports
~~~~~~~~~~~~~~~~~~

Passports are introduced by the authenticator system. A passport contains
all information that should be used for authentication. This object is used

> to check authentication by the security system.

Reviewer 4: Hmm, these 3 sentences don't really *say* much :). Maybe just:

> A passport is an object that contains the user that will be authenticated as well
> as other pieces of information, like whether a password should be checked or
> if "remember me" functionality should be enabled.
